internal/formats/common/spdxhelpers/to_syft_model.go

```diff
@@ -0,0 +1,373 @@
+)
+
+func ToSyftModel(doc *spdx.Document2_2) (*sbom.SBOM, error) {
+	spdxIDMap := make(map[string]interface{})
+
+	collectSyftPackages(spdxIDMap, doc)
+
+	collectSyftFiles(spdxIDMap, doc)
+
+	catalog := pkg.NewCatalog()
+
+	for _, v := range spdxIDMap {
+		if p, ok := v.(*pkg.Package); ok {
+			catalog.Add(*p)
+		}
+	}
+
+	return &sbom.SBOM{
+		Artifacts: sbom.Artifacts{
+			PackageCatalog:    catalog,
+			LinuxDistribution: findSyftLinuxRelease(doc),
+		},
+		Relationships: toSyftRelationships(spdxIDMap, doc),
+	}, nil
+}
+
+func findLinuxReleaseByDocument(doc *spdx.Document2_2) *linux.Release {
+	var p *spdx.Package2_2
+
+	if r := findSpdxRelationshipByType(doc, "DESCRIBES"); r != nil {
+		if string(r.RefA.ElementRefID) == "DOCUMENT" {
+			p = findSpdxPackageByID(doc, r.RefB.ElementRefID)
+		}
+	}
+
+	if r := findSpdxRelationshipByType(doc, "DESCRIBED_BY"); r != nil {
+		if string(r.RefB.ElementRefID) == "DOCUMENT" {
+			p = findSpdxPackageByID(doc, r.RefA.ElementRefID)
+		}
+	}
+
+	var name string
+	var version string
+
+	if p != nil {
+		name = p.PackageName
+		version = p.PackageVersion
+	}
+
+	if p != nil && (name == "" || version == "") {
+		// SPDX has relationship: PACKAGE_OF
+		// and Package External Reference PACKAGE-MANAGER
+		// spec references "PACKAGE-MANAGER" but JSON schema has "PACKAGE_MANAGER"
+		ref := findSpdxReferenceByName(p, "PACKAGE-MANAGER", "PACKAGE_MANAGER")
+
+		if ref != nil {
+			parts := strings.Split(ref.Locator, "-")
+			if len(parts) > 1 {
+				name = parts[0]
+				version = parts[1]
+			}
+		}
+	}
+
+	if name != "" && version != "" {
+		return &linux.Release{
+			PrettyName: name,
+			Name:       name,
+			ID:         name,
+			IDLike:     []string{name},
+			Version:    version,
+			VersionID:  version,
+		}
+	}
+
+	return nil
+}
+
+func findLinuxReleaseByPURL(doc *spdx.Document2_2) *linux.Release {
+	for _, p := range doc.Packages {
+		purlValue := extractPURL(p)
+		if purlValue != "" {
+			purl, err := packageurl.FromString(purlValue)
+			if err != nil {
+				log.Warnf("unable to parse purl: %s", purlValue)
+			} else {
+				for _, qualifier := range purl.Qualifiers {
+					if qualifier.Key == "distro" {
+						parts := strings.Split(qualifier.Value, "-")
+						if len(parts) > 1 {
+							name := parts[0]
+							version := parts[1]
+							return &linux.Release{
+								PrettyName: name,
+								Name:       name,
+								ID:         name,
+								IDLike:     []string{name},
+								Version:    version,
+								VersionID:  version,
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func findSyftLinuxRelease(doc *spdx.Document2_2) *linux.Release {
+	r := findLinuxReleaseByDocument(doc)
+	if r != nil {
+		return r
+	}
+
+	r = findLinuxReleaseByPURL(doc)
+	if r != nil {
+		return r
+	}
+
+	return nil
+}
+
+func findSpdxRelationshipByType(doc *spdx.Document2_2, typ string) *spdx.Relationship2_2 {
+	for _, r := range doc.Relationships {
+		if typ == r.Relationship {
```

Comment:
this utility function looks like it will only return the first instance of a matching relationship type, however, there may be several instances of a type

Reply:
Ended up deleting this after addressing some other feedback 🎉